rfcs/2021-05-17-7469-load-balancing-aggregators.md

```diff
@@ -0,0 +1,99 @@
+* [Outstanding Questions](#outstanding-questions)
+* [Plan Of Attack](#plan-of-attack)
+
+## Scope
+
+Load balancing will be a concern for any `sink` or `source` supported by Vector; some of which can use a general solution (ex: load balancing for our HTTP based `sinks`) and some of which are specific to the component (ex: Kafka or Elasticsearch `sinks`). Due to the breadth of the topic, this RFC will focus on three specific cases for load balancing while also giving consideration to future adoption for other components.
+
+* Vector agent to Vector aggregator
+* Datadog agent to Vector aggregator
+* Syslog agent to Vector aggregator
+
+## Motivation
+
+Today Vector lacks the capacity to scale horizontally (by increasing replicas) when deployed as an aggregator. This limits Vector aggregators in both reliability and performance, causing adoption concerns for users. A single aggregator will be limited in performance by the resources that can be dedicated to it, presumably with some (currently) unknown upper bounds. Vector aims to be vendor neutral, and as such we should provide the capacity to scale and load balance across Vector aggregators regardless of environment or upstream event collectors.
+
+## Internal Proposal
+
+Include a configuration for a dedicated reverse proxy that will be deployed as part of the vector-aggregator Helm chart, as well as documented configuration and installation instructions for users that run outside of Kubernetes. We should provide basic, but functional, configurations out-of-the box to enable users to "one click" install Vector as an aggregator. The proxy should dynamically resolve downsteam Vector instances but allow for consistent targets in situations that require it (aggregation transforms).
+
+Below is a basic HAProxy configuration configured to leverage service discovery in a Kubernetes cluster:
+
+```haproxy
+resolvers coredns
+    nameserver dns1 kube-dns.kube-system.svc.cluster.local:53
+    hold timeout         600s
+    hold refused         600s 
+frontend vector
+    bind *:9000
+    default_backend vector_template
+backend vector_template
+    balance roundrobin
+    option tcp-check         
+    server-template srv 10 _vector._tcp.vector-aggregator-headless.vector.svc.cluster.local resolvers coredns check
+```
+
+## Rationale
+
+* Configuring an external reverse proxy for load balancing allows for load balancing regardless of the upstream agent.
+* Using a dedicated reverse proxy to load balance requests for Vector aggregators should support the largest spread of `sources` with the smallest amount of engineering effort.
+* A solution outside of Vector itself ensures that users can reliably adopt Vector as an aggregator without replacing their existing infrastructure.
+* Most organizations are likely familiar with operating _some_ class of reverse proxy.
+* Dedicated reverse proxy can be specialized and optimized for its task, and the same can be said for Vector itself.
```

Comment:
I buy this rationale. I'm wary about adding additional functionality into vector -- a good reverse proxy is not a small chunk of code -- and its already a crowded space with solid options available. While I do think we should gear up our helm chart to ship with one reverse proxy, I foresee that we're going to need to support / test with a handful of popular options.

Reply:
I think it's fine to start with a single "blessed" proxy, and expand from there based on demand. I'm afraid of how large the testing matrix is going to get, considering kubernetes versions, components, proxies, etc

Reply:
Yep, yep. It can get out of control fast. My preference would be to start with just the "blessed" things and then lump in more options as users need them in a compelling way.